Remove object-hash, fixes #42

mazondo opened the pull request:
Instead of just fixing the import, decided to try and remove it all together.  We're only hashing objects, so the other support from object-hash is overkill.  since we already load md5 for gravatar, I used that instead of sha1, not really much collision potential here and pretty minimal consequences if it happens.

Cuts about 35k off the final uglified source.

mazondo commented:
Need to make some quick changes to this, there's some cases where fastboot errors still pop up

mazondo commented:
It looks like a lot of the issue has to do with moving from the service to the initializer.  Fastboot handles initializers differently so I'm guessing that's what we're hitting.  Is there a reason it was moved to a util from a service @Exelord.  It seems like after initialization it's basically acting as a service anyways?

Exelord commented on addon/utils/store.js:
```diff
@@ -21,11 +21,30 @@ export default EmberObject.extend({
+
+  _ensureHashable(properties) {
+    if (typeOf(properties) !== 'object') {
+      // this should never happen, but just in case
```
Actually, it can happen because the value of the properties key can be a whatever

mazondo replied:
👍 I'll add some test cases to cover it.

Exelord commented on addon/utils/store.js:
```diff
@@ -21,11 +21,30 @@ export default EmberObject.extend({
   },
 
   initialsFor(properties) {
-    let key = hash(properties);
```
Can we just write a polyfill on that function without changing the public interface implementation?
I mean, instead of `_ensureHashable` just write our own `_hash` method which will replace the object-hash module.

mazondo replied:
Sure, np.

Exelord commented on addon/utils/store.js:
```diff
@@ -21,11 +21,30 @@ export default EmberObject.extend({
+      return properties.toString();
+    }
+
+    let pKeys = Object.keys(properties).sort();
```
Just curious, why we are sorting it?

mazondo replied:
@Exelord Just an attempt at increasing the cache hit rate by standardizing key order.

{a: 'b', c: 'd'} should equal {c: 'd', a: 'b'}.

Exelord replied:
Right! I will also add tests for that case.

Exelord commented on addon/utils/store.js:
```diff
@@ -21,11 +21,30 @@ export default EmberObject.extend({
+
+    let pKeys = Object.keys(properties).sort();
+    let result = [];
+    pKeys.forEach((key) => {
```
What about `.map` here?

mazondo replied:
I didn't do a `.map` because I wasn't positive how the nested check would work in that case.  I didn't want to end up with `[a, b, [c, d]]` so chose to concat instead.  Open to that being incorrect

Exelord commented on addon/utils/store.js:
```diff
@@ -21,11 +21,30 @@ export default EmberObject.extend({
+    let result = [];
+    pKeys.forEach((key) => {
+      let value = properties[key];
+      if (typeOf(value) !== 'string') {
```
we should definitely have more tests for every of the possible case with string number object array function and custom class

Exelord replied:
I will take care of that

Exelord replied:
Done in #44

mazondo replied:
sure, np

Exelord commented on package.json:
```diff
@@ -65,15 +65,13 @@
     "broccoli-merge-trees": "2.0.0",
-    "ember-cli-babel": "6.11.0",
-    "object-hash": "1.2.0"
+    "blueimp-md5": "2.10.0"
```
this dep should be at the beginning. Alphabetical order.

Exelord commented:
Hey! Thanks for your work. I definitely agree that removing object-hash in favor of our simple implementation will be a nice improvement. Please, keep working on that, let's close it! :)

But is that fix the cause of #42? How did you reproduce the issue? It doesn't happen in demo development/production env with fastboot

mazondo commented:
@Exelord Thanks for the review, I have some changes coming shortly. The real issue for #42 is actually the initializer, so I can do a smaller PR that fixes the immediate concern and then look at removing object-hash in this one.

Unfortunately, the fix for the fastboot issue is to return the store to a service instead of initializing it.  The fastboot issue is only evident in production and seems to be while using fastboot-app-server.

Exelord commented:
The reverting of the initializer is not a fix. There is v3.5.0 released which will work perfectly as a quick fix with all the upgrades.

I need to be able to reproduce that error in case of further investigation because it should work normally. Did you find a way to catch the error?

Exelord commented:
Just a rough idea,
What if would replace the object hash with https://github.com/yeoman/stringify-object and use md5 on it? It looks like we are trying to achieve the same result.

mazondo commented:
Added the tests you asked for, also changed to a sep util for hash to maintain the internal api like you mentioned.

Tests are passing on this one, but it does revert the service.

I hear you on it not being the fix you're looking for, but candidly I'm not convinced on why you're recreating the concept of a service internally here, so I'm proposing it as an option.

I can work on getting you a test app, but probably not until later in the week.  The gist tho is that if you run your app with fastboot-app-server, you'll hit it.  Ember apps don't run with just fastboot in production, it's not intended to be a production server, so fastboot-app-server is the suggested solution.

mazondo commented:
My concern, btw, just comes from not leveraging ember internals when possible.  FastBoot in production is a single long-running app process and is prone to memory leaks and issues of that sort as a result.  Initializers tend to be prone to those types of issues.

We could definitely look at other stringify options, however I do think that the use case here is fairly small and they may be a bit overkill for the goal.

mazondo commented:
Respectfully man, having used fastboot in production for almost a year now, we're not comfortable with that approach so we'll probably stick with our fork that utilizes the service.  I realize singletons are not considered internals, but given the long running nature of ember fastboot, you're just introducing too many potential issues needlessly.

I can definitely add tests for handling nested order, though again I do question the utility of this for generating cache keys.  I think my push back would be on what's required to generate a unique key and at what point you lose value given the tradeoff of library size in a single-page-app environment.

For example, even the md5 seems relatively overkill to me but wanted my PR to match the spirit of your original.  Our fork of this library in production is 3k, a full 35k lower than your current master.  It's a pretty big win, I'd really encourage continuing to think in that direction and considering the tradeoffs.

Exelord commented:
Thank you for your work! The cache store had been fixed as well object-hash dependency had been removed.

Im agree that current store method is not perfect, but it works pretty well. If you have an idea how to improve it, go on.
